File: LetterCombination/LetterCombinationv1.java
package LetterCombination;

import java.util.*;

public class LetterCombinationv1 {
    public static void main(String[] args){
        System.out.println(letterCombinations("23"));
        System.out.println(letterCombinations("2"));
        System.out.println(letterCombinations(""));
    }

    public static List<String> letterCombinations(String digits){
        List<String> result = new ArrayList<>();
        if (digits.length() == 0) {return result;}
        if (digits.length() == 1) {return Arrays.asList(getLetters(digits.charAt(0)));}
        findCombinations(digits, 0, new StringBuilder(), result);
        return result;
    }

    public static void findCombinations(String digits, int index, StringBuilder combination, List<String> result){
        if (index == digits.length()){
            result.add(combination.toString());
            return;
        }
        String letters = getLetters(digits.charAt(index));
        for (int i = 0; i < letters.length(); i++){
            combination.append(letters.charAt(i));
            findCombinations(digits, index + 1, combination, result);
            combination.deleteCharAt(combination.length() - 1);
        }
    }
    public static String getLetters(char letter){
        switch (letter) {
            case '2': return "abc";
            case '3': return "def";
            case '4': return "ghi";
            case '5': return "jkl";
            case '6': return "mno";
            case '7': return "pqrs";
            case '8': return "tuv";
            case '9': return "wxyz";
            default: return "";
        }
    }
}

File: LetterCombination/LetterCombinationv1.py
def letterCombinations(digits):
    result = []
    if len(digits) == 0:
        return result
    findCombinations(digits, 0, "", result)
    return result

def findCombinations(digits, index, combination, result):
    if index == len(digits):
        result.append(combination)
        return
    letters = getLetters(digits[index])
    for ch in letters:
        findCombinations(digits, index + 1, combination + ch, result)

def getLetters(letter):
    mapping = {
        '2': "abc", '3': "def", '4': "ghi",
        '5': "jkl", '6': "mno", '7': "pqrs",
        '8': "tuv", '9': "wxyz"
    }
    return mapping.get(letter, "")

print(f"For {23}, the solution is {letterCombinations("23")}")
print(f"For {2}, the solution is {letterCombinations("2")}")
